packages/starlight/integrations/expressive-code/index.ts

```diff
@@ -0,0 +1,250 @@
+import fs from 'node:fs';
+import path from 'node:path';
+import {
+	astroExpressiveCode,
+	ExpressiveCodeTheme,
+	type AstroExpressiveCodeOptions,
+	pluginFramesTexts,
+	addClassName,
+} from 'astro-expressive-code';
+import type { AstroIntegration } from 'astro';
+import type { StarlightConfig } from '../../types';
+import type { createTranslationSystemFromFs } from '../../utils/translations-fs';
+
+export * from 'astro-expressive-code';
+
+export type StarlightExpressiveCodeOptions = Omit<AstroExpressiveCodeOptions, 'themes'> & {
+	/**
+	 * The color themes that should be available for your code blocks.
+	 *
+	 * CSS variables will be generated for all themes, allowing to select the theme to display
+	 * using CSS. If you specify one dark and one light theme, the appropriate theme matching
+	 * Starlight's dark mode switch state will be displayed automatically. Have a look at the
+	 * `useStarlightDarkModeSwitch` option for more details.
+	 *
+	 * The following item types are supported in this array:
+	 * - any theme name bundled with Shiki (e.g. `dracula`)
+	 * - any theme object compatible with VS Code or Shiki (e.g. imported from an NPM theme package)
+	 * - any ExpressiveCodeTheme instance (e.g. using `ExpressiveCodeTheme.fromJSONString(...)`
+	 *   to load a custom JSON/JSONC theme file yourself)
+	 *
+	 * Defaults to the dark and light variants of the "Night Owl" theme by Sarah Drasner,
+	 * which are also exported from `@astrojs/starlight/expressive-code` as
+	 * `getStarlightDefaultThemes`.
+	 */
+	themes?: AstroExpressiveCodeOptions['themes'];
+	/**
+	 * Determines if CSS code should be added to the site that automatically displays
+	 * dark or light code blocks depending on Starlight's dark mode switch state.
+	 *
+	 * Defaults to `true` if the `themes` option is not set or contains at least
+	 * one dark and one light theme, and `false` otherwise.
+	 */
+	useStarlightDarkModeSwitch?: boolean | undefined;
+	/**
+	 * Determines if Starlight's CSS variables should be used for the colors of UI elements
+	 * (backgrounds, buttons, shadows etc.) instead of the colors provided by themes.
+	 *
+	 * Defaults to `true` if the `themes` option is not set (= you are using the default themes),
+	 * and `false` otherwise.
+	 */
+	useStarlightUiThemeColors?: boolean | undefined;
+};
+
+export const starlightExpressiveCode = (
+	opts: StarlightConfig,
+	useTranslations: ReturnType<typeof createTranslationSystemFromFs>
+): AstroIntegration[] => {
+	const { locales, defaultLocale, expressiveCode } = opts;
+	if (expressiveCode === false) return [];
+	const config: StarlightExpressiveCodeOptions =
+		typeof expressiveCode === 'object' ? expressiveCode : {};
+
+	const {
+		themes = [],
+		customizeTheme,
+		styleOverrides: { textMarkers: textMarkersStyleOverrides, ...otherStyleOverrides } = {},
+		useStarlightDarkModeSwitch,
+		useStarlightUiThemeColors = config.themes === undefined,
```

Comment:
Coming back with one more bit of feedback as I work on the docs.

While I understand the idea behind making the `useStarlightDarkModeSwitch` and `useStarlightUiThemeColors` options responsive to how `themes` is configured — basically trying to guess what a user wants based on `themes` — I actually think introducing this interdependence between configuration options can be quite confusing. Instead of having a static default, a user has to work out what is going on in another config key and then understand the mapping.

I think I’d like to propose making `useStarlightUiThemeColors` always default to `true`, letting people override it with `false` when that’s not what they want:

```suggestion
		useStarlightDarkModeSwitch,
		useStarlightUiThemeColors = true,
```

`useStarlightDarkModeSwitch` is obviously more heavily dependent on `themes` — if there’s only one theme, or no theme for one of light/dark, we obviously can’t sync with the dark mode. So I think the current behaviour is good. I think I can make the docs clear in this case by saying the default is `true` and then note the exception that if you set a custom theme and don’t provide both dark and light the setting can’t apply. Still drafting this, but something like:

> #### `useStarlightDarkModeSwitch`
>
> **type:** `boolean`  
> **default:** `true`
>
> When `true`, code blocks automatically switch between light and dark themes when the site theme changes.
>
> :::note
> When setting `themes`, you must provide at least one dark and one light theme for the Starlight dark mode switch to work.
> :::

Reply:
It's certainly a valid point that interdependent config options are more difficult to understand. My idea was that if users explicitly specify their own set of themes instead of just keeping the defaults, they would likely expect the themes to look like they do in the VS Code theme gallery, and not like a hybrid between the theme and Starlight.

I could actually imagine dropping this option altogether if we add the `starlight-light` and `starlight-dark` string options as discussed. That the theme respects the Starlight colors would then be a feature of the Starlight themes. What do you think?

Reply:
I think the option has very valid use cases so I wouldn’t drop it. It lets you choose a syntax highlighting theme without also opting out of the UI frame fitting with your wider site theme.

For example, here’s a quick example of `github-dark` with its built-in colours and with the site colours:

| `useStarlightUiThemeColors` unset | `useStarlightUiThemeColors: true` |
|---|---|
| <img width="583" alt="image" src="https://github.com/withastro/starlight/assets/357379/3013231b-fc98-403f-8dd3-f5cb99c18e0d"> | <img width="587" alt="image" src="https://github.com/withastro/starlight/assets/357379/cbf00c3f-e6ef-42a2-a27e-7f00854a3e31"> |

Not super clear without the full site context, but I think having the UI harmonious with the broader site is actually a pretty good default and opting into using the theme for the full UI would be the add-on.

Can be shown even with the default theme where I find disabling the Starlight colours pretty odd (again maybe more out of place with the full page context, but keeping the screenshots small for GitHub):

| `useStarlightUiThemeColors` unset | `useStarlightUiThemeColors: false` |
|---|---|
| <img width="586" alt="image" src="https://github.com/withastro/starlight/assets/357379/71dec446-0c37-49a3-980a-6ebfb23b88da"> | <img width="587" alt="image" src="https://github.com/withastro/starlight/assets/357379/2883b9d6-8981-4fb2-8b76-4e18816aafa1"> |

So I’d definitely keep the option, but I think making it a static default of `true` makes most sense to me.

Reply:
Ok, I'll make it default to `true` then. I guess if users notice that their chosen themes look different in Starlight than what they expected, they will probably find this option. :)

Reply:
After making the suggested change to the code and always defaulting the option to `true`, I remembered why I made it dependent on the themes in the first place: If you set `themes` to a single theme like `['github-dark']`, and then switch your Starlight theme to light, Starlight's CSS variables will switch from dark to light without switching the syntax highlighting theme (because there is only one), which results in a dark theme displayed with a light background color.

My automatic color contrast correction cannot fix this scenario because it works per theme and has to rely on the background having a certain lightness. If this lightness changes dramatically at runtime, there is no way to preserve color contrast anymore.

Reply:
Still thinking about this issue and I don't see any sensible solutions that do not depend on the `themes` option.

For the sake of finalizing this PR, I'd suggest keeping my initial proposal of the option automatically defaulting to false when `themes` has been customized. I would not want to release default settings that can lead to inaccessible color contrast.

Reply:
Ahhhhhhh, right. One hypothetical option would I guess to be to use the Starlight background color but pin it to the syntax theme’s light/dark type. But that’s not straightforward because you only have access to the global CSS variable which will change.

Yes, I think we should probably revert back to your original choice. Sorry for the thrash on that — I didn’t test this case while trying things out yesterday.

Reply:
No worries at all! We're both striving for simplicity and ease of use, and it's absolutely valid to consider alternatives. Discussing them helps us all develop a better understanding of the options. I'll keep this behavior unchanged then.

Do you want to adapt the description in the docs or should I try to fit your tone in the updated version?

Reply:
Up to you. Maybe as a first step: do YOU want to go through accepting Sarah’s suggestions or would you prefer I do that?

Reply:
I think you're better suited to get the language details right as a native speaker, and as the Starlight lead you bear the burden of making all contributions fit together nicely, so I'd gladly leave this to you if you have the time. :)